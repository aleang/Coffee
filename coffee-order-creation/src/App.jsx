import React, { useEffect, useState, useRef } from "react";
import './App.css';

const orderExample = `0900 1730
2022-03-17 10:17:06 001
2022-03-21 09:00 2
2022-03-16 12:34:56 002
2022-03-21 09:00 2`;

function App() {
  const [order, setOrder] = useState('');
  const [serverResponse, setResponse] = useState(null);
  const [isLoading, setLoading] = useState(false);
  const coffeeOrderInput = useRef();

  useEffect(() => {
    if (!isLoading) return;

    fetch("http://localhost:8080/coffee-order-creation", {
      method: 'POST',
      mode: 'cors',
      headers: new Headers({
        "Content-Type": "text/plain",
        "Accept": "text/plain"
      }),
      body: order
    })
      .then((res) => res.json())
      .then((data) => {
        if (data.error) throw new Error(data.error)
        setResponse(data.deliverySchedule)
        setLoading(false)
      })
      .catch((error) => {
        setResponse(error.message)
        setLoading(false)
      });
  }, [order, isLoading]);

  const placeOrder = () => {
    const orderTextareaDom = document.getElementById("orderTextarea")
    setOrder(orderTextareaDom.value.trim())
    setLoading(true)
    setResponse("")
  }

  const dismissOutput = () => {
    setResponse(null);
    coffeeOrderInput.current.focus();
  }

  return (
    <div className="App">
      <header className="App-header">
        <h1>☕</h1>
        <div className={`place-order ${!!serverResponse ? "shortest" : ""} ${isLoading ? "shorter" : ""} `}>
          <p>Place coffee orders</p>
          
          <textarea id="orderTextarea" placeholder={orderExample} data-testid="coffee-order-input" ref={coffeeOrderInput}>
          </textarea>
  
          <input type="submit" value="Place Order" onClick={() => placeOrder()} data-testid="submit-button" />
        
        </div>
        { isLoading && <span className="loading" data-testid="loading-icon">⌛</span>}
        { serverResponse && 
          <pre className="after-order" >
            <span data-testid="coffee-order-output">{serverResponse}</span>
            <span className="close" onClick={() => { dismissOutput()}} data-testid="dismiss-output">x</span>
          </pre>
        }

      </header>
    </div>
  );
}

export default App;
