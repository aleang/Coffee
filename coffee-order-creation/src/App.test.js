import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import App from './App'
import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils'

describe('Renders the app', () => {

  const mockFetch = (response, isReject = false) => {
    if (isReject) {
      global.fetch = jest.fn(response)
    } else {
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: response,
        })
      )
    }
  }

  afterEach(() => {
    jest.clearAllMocks()
  })
  
  test('renders App as per snapshot', () => {
    const container = render(<App />)
    expect(container).toMatchSnapshot()
  })

  test('accepts input and show loading icon', async () => {
    mockFetch(() => Promise.resolve({ deliverySchedule: '12:34 12:45 001' }))

    act(() => {
      render(<App />)
    })

    userEvent.type(screen.getByTestId('coffee-order-input'), 'A')  
    fireEvent.click(screen.getByTestId('submit-button'))
    expect(screen.getByTestId('loading-icon')).toBeInTheDocument()
    await waitFor(() => {
      expect(screen.queryByTestId('loading-icon')).not.toBeInTheDocument()
    })
  })

  test('dismiss output and give focus back to Order input', async () => {
    mockFetch(() => Promise.resolve({ deliverySchedule: '12:34 12:45 001' }))
    
    act(() => {
      render(<App />)
    })

    userEvent.type(screen.getByTestId('coffee-order-input'), 'A')  
    fireEvent.click(screen.getByTestId('submit-button'))
    await waitFor(() => {
      expect(screen.queryByTestId('loading-icon')).not.toBeInTheDocument()
      expect(screen.getByTestId('coffee-order-output').textContent).toBe('12:34 12:45 001')
      fireEvent.click(screen.getByTestId('dismiss-output'))
      expect(screen.getByTestId('coffee-order-input')).toHaveFocus()
    })
  })
  
  test('accepts input and handle error', async () => {
    mockFetch(() => Promise.reject(new Error('bad input')), true)

    act(() => {
      render(<App />)
    })

    userEvent.type(screen.getByTestId('coffee-order-input'), 'A')  
    fireEvent.click(screen.getByTestId('submit-button'))
    expect(screen.getByTestId('loading-icon')).toBeInTheDocument()
    await waitFor(() => {
      expect(screen.getByTestId('coffee-order-output').textContent).toBe('bad input')
    })
  })

  test('accepts input and handle response with "error"', async () => {
    mockFetch(() => Promise.resolve({ error: 'terrible input' }))

    act(() => {
      render(<App />)
    })

    userEvent.type(screen.getByTestId('coffee-order-input'), 'A')
    fireEvent.click(screen.getByTestId('submit-button'))
    await waitFor(() => {
      expect(screen.queryByTestId('loading-icon')).not.toBeInTheDocument()
      expect(screen.getByTestId('coffee-order-output').textContent).toBe('terrible input')
    })
  })
})