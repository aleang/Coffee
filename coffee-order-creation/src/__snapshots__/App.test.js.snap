// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the app renders App as per snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="App-header"
        >
          <h1>
            ☕
          </h1>
          <div
            class="place-order   "
          >
            <p>
              Place coffee orders
            </p>
            <textarea
              data-testid="coffee-order-input"
              id="orderTextarea"
              placeholder="0900 1730
2022-03-17 10:17:06 001
2022-03-21 09:00 2
2022-03-16 12:34:56 002
2022-03-21 09:00 2"
            />
            <input
              data-testid="submit-button"
              type="submit"
              value="Place Order"
            />
          </div>
        </header>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="App-header"
      >
        <h1>
          ☕
        </h1>
        <div
          class="place-order   "
        >
          <p>
            Place coffee orders
          </p>
          <textarea
            data-testid="coffee-order-input"
            id="orderTextarea"
            placeholder="0900 1730
2022-03-17 10:17:06 001
2022-03-21 09:00 2
2022-03-16 12:34:56 002
2022-03-21 09:00 2"
          />
          <input
            data-testid="submit-button"
            type="submit"
            value="Place Order"
          />
        </div>
      </header>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
